[CORE-13950] Fix L1 reader gap skipping

dotnwat opened the pull request:
* Fixes the L1 reader gap skipping issue
* Removes all of the explicit states from the L1 reader. Now the state machine is implicit in the structure of the main reader coroutine.
* Reenables reconciler in RNOT tests
* Adds a test_suite_cloud_topics.yml to quickly run cloud topics enabled tests

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

* none

dotnwat committed:
test: reenable cloud topics reconciler in RNOT

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
ct: use implicit state machine in L1 reader

This commit removes the explicit states from the L1 reader and most of
the shared state (e.g. current object, current batches, ...) and encodes
the state machine directly in the coroutine logic.

This commit also addresses CORE-13950 which relates to the L1 reader
skipping offsets leading to gaps that produce incorrect fetch behavior.

The problem before was that reader would attempt to move the next offset
forward if no batches were read. However, the code which moved the
offset forward could not distinguish between no batches because the
reader read a full object vs no batches because the reader reach its
byte limit. In the later case the reader would increment the offset and
end up skipping batches when it should have returned what it had read
and the marked itself as end of stream.

This commit handles that case by identifying the over the limit scenario
and explicitly marking the reader as being at end of stream.

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
ct: rename current_object to object_info

Now that we don't have a "current" object in the reader we rename the
type to better reflect reality.

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
ct: simplify next offset handling in l1 reader

Simplification includes the observation that the object optional is
never disengaged so that removes an entire branch from the calculation.

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat commented on src/v/cloud_topics/level_one/frontend_reader/level_one_reader.cc:
```diff
@@ -60,76 +60,49 @@ ss::future<model::record_batch_reader::storage_t>
-            _batches = co_await materialize_batches_from_object_offset(
-              _current_obj.value(), _next_offset, deadline);
-            _state = state::materialized;
-            [[fallthrough]];
-        case state::materialized:
-        case state::end_of_stream:
-            // Handled in the next switch statement.
-            break;
+
+        auto batches = co_await materialize_batches_from_object_offset(
+          object.value(), _next_offset, deadline);
```
Yeh. I'll add this and a few other optimizations once I see that tests are passing.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--104cb548-5767-4475-b293-564b6b02ae6c-->
<!--TEST RESULTS-->
<!--4fc6e10d-0439-45e1-b550-af5ca2fe7f38-->
<!--TEST RESULTS-->
<!--c7ada334-9d7e-44bd-be72-32e914acdfda-->
<!--TEST RESULTS-->
<!--6dc0c9f1-2f61-4089-9668-2939e0591074-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74470">build#74470</a></summary>


| test_class                      | test_method                    | test_arguments                                                               | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                               |
|:--------------------------------|:-------------------------------|:-----------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests   | test_replication_with_failures | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/74470#0199f5af-5a3e-48cc-a371-cfb849ff6761 | FLAKY         | 6/21     | upstream reliability is '75.27405602923264'. current run reliability is '28.57142857142857'. drift is 46.70263 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_with_failures |
| Datalake3rdPartyMaintenanceTest | test_e2e_basic                 | {"catalog_type": "nessie", "cloud_storage_type": 1, "query_engine": "spark"} | integration | https://buildkite.com/redpanda/redpanda/builds/74470#0199f5b5-222a-4952-8d7f-59ea4165ca7b | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=Datalake3rdPartyMaintenanceTest&test_method=test_e2e_basic               |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74481">build#74481</a></summary>


| test_class                    | test_method                    | test_arguments                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                               |
|:------------------------------|:-------------------------------|:----------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| AuditLogTestKafkaApi          | test_no_auth_enabled           | {"audit_transport_mode": "rpc"}                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/74481#0199f8da-86b5-42c4-94c4-c067d069150b | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=AuditLogTestKafkaApi&test_method=test_no_auth_enabled                    |
| ShadowLinkingReplicationTests | test_replication_basic         | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "redpanda"}} | integration | https://buildkite.com/redpanda/redpanda/builds/74481#0199f8da-86b5-42c4-94c4-c067d069150b | FLAKY         | 16/21    | upstream reliability is '94.88636363636364'. current run reliability is '76.19047619047619'. drift is 18.69589 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic         |
| ShadowLinkingReplicationTests | test_replication_with_failures | null                                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/74481#0199f8d8-b909-4523-b2f1-0598d0b10c44 | FLAKY         | 7/21     | upstream reliability is '71.2253829321663'. current run reliability is '33.33333333333333'. drift is 37.89205 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_with_failures |
| ShadowLinkingReplicationTests | test_replication_with_failures | null                                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/74481#0199f8da-86b6-4f21-a392-d0f8c659ec4f | FLAKY         | 12/21    | upstream reliability is '71.2253829321663'. current run reliability is '57.14285714285714'. drift is 14.08253 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_with_failures |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74491">build#74491</a></summary>


| test_class               | test_method                                             | test_arguments                               | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                                                   |
|:-------------------------|:--------------------------------------------------------|:---------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| NodesDecommissioningTest | test_decommissioning_finishes_after_manual_cancellation | {"cloud_topic": false, "delete_topic": true} | integration | https://buildkite.com/redpanda/redpanda/builds/74491#0199fa10-df78-45fa-82ce-861c05b6a3b6 | FLAKY         | 21/22    | upstream reliability is '97.53694581280789'. current run reliability is '95.23809523809523'. drift is 2.29885 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_finishes_after_manual_cancellation |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74496">build#74496</a></summary>


| test_class                    | test_method                          | test_arguments                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                   | test_history                                                                                                                                                               |
|:------------------------------|:-------------------------------------|:---------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| LeaderEpochTest               | TestGetLeaderEpochWhileReconciling   |                                                                                                                | unit        | https://buildkite.com/redpanda/redpanda/builds/74496#0199fdce-c4e2-4115-b4a9-c6681b03f724 | FAIL          | 0/1      |                                                                                                                                                                          |                                                                                                                                                                            |
| AuditLogTestKafkaApi          | test_no_auth_enabled                 | {"audit_transport_mode": "rpc"}                                                                                | integration | https://buildkite.com/redpanda/redpanda/builds/74496#0199fdeb-dd69-4540-9981-9c814b1a8f0a | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS                | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=AuditLogTestKafkaApi&test_method=test_no_auth_enabled                    |
| ShadowLinkBasicTests          | test_topic_properties_replicated     | {"source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}} | integration | https://buildkite.com/redpanda/redpanda/builds/74496#0199fdeb-dd6a-424f-9d3d-9df38a86e236 | FLAKY         | 17/21    | upstream reliability is '95.76271186440678'. current run reliability is '80.95238095238095'. drift is 14.81033 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkBasicTests&test_method=test_topic_properties_replicated        |
| ShadowLinkingReplicationTests | test_replication_with_failures       | null                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/74496#0199fdeb-dd6a-424f-9d3d-9df38a86e236 | FLAKY         | 7/21     | upstream reliability is '70.58823529411765'. current run reliability is '33.33333333333333'. drift is 37.2549 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_with_failures |
| ShadowLinkingReplicationTests | test_replication_with_failures       | null                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/74496#0199fe52-b0f5-48fa-b81e-029c413794c1 | FLAKY         | 5/21     | upstream reliability is '70.58823529411765'. current run reliability is '23.809523809523807'. drift is 46.77871 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_with_failures |
| DataMigrationsApiTest         | test_creating_and_listing_migrations | null                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/74496#0199fdeb-dd67-49b4-ba1b-4b7b57dfadf9 | FLAKY         | 19/21    | upstream reliability is '99.79919678714859'. current run reliability is '90.47619047619048'. drift is 9.32301 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_creating_and_listing_migrations   |
</details>
<!--/body-->

dotnwat committed:
ct: pass object reference

This generally preferred over passing references to smart pointers.

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
ct: advance l1 reader next offset automatically

Prior to this commit we relied upon the record batch reader
infrasturcture to request new slices to drive next offset forward when
previous requests did not return batches. This commit introduces an
inner loop so that the reader does not rely on record batch reader to
drive that progress.

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
ct: move state change checks to top-level

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
test: add test suite for cloud topics enabled tests

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
ct: make function name more descriptive

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat commented:
/ci-repeat 1
skip-redpanda-build
skip-rebase
tests/rptest/test_suite_cloud_topics.yml

dotnwat commented:
/ci-repeat 2
dt-repeat=2
dt-nodes=2
skip-redpanda-build
skip-rebase
tests/rptest/test_suite_cloud_topics.yml

rockwotj commented on src/v/cloud_topics/level_one/frontend_reader/level_one_reader.cc:
```diff
@@ -60,91 +60,66 @@ ss::future<model::record_batch_reader::storage_t>
+ss::future<model::record_batch_reader::storage_t>
+level_one_log_reader_impl::read_some(
+  model::timeout_clock::time_point deadline) {
+    while (true) {
```
Should this be while not end of stream?

dotnwat replied:
from a style perspective or optimization/correctness? if the former, i can change it i have no preference. if the later, I don't see an issue: load_slice is not invoked by record batch reader if is_end_of_stream is true, and the operations following the is_end_of_stream check in this will loop do not set end_of_stream.

rockwotj commented on src/v/cloud_topics/level_one/frontend_reader/level_one_reader.cc:
```diff
@@ -60,91 +60,66 @@ ss::future<model::record_batch_reader::storage_t>
+              "stream",
+              _next_offset,
+              _config.max_offset);
+            set_end_of_stream();
```
This was silently introduced, but I am glad you added it for consistency

dotnwat replied:
do you mean `set_end_of_stream`? i alluded to it in the commit message of the 3rd (or 4th commit), but yeh I didn't call it out explicitly:

```
This commit [...]
and explicitly marking the reader as being at end of stream.
```

rockwotj reviewed (commented):
Looks great! I would love to see a unit test added since this did fix a bug.

dotnwat committed:
ct: add reader gap test for CORE-13950

Adds test that is expected to fail prior to applying the subsequent
commits which fix CORE-13950. In short, the L1 reader will read gaps in
error (i.e. not due to compaction) in the following case:

1. speculative byte limit is exceeded in reader::read_batches
2. actual (not speculative) byte limit is not exceeded
3. (above causes reader to advance next offset and continue reading)
4. A small batch is encountered as the first batch in an L1 object which
   is small enough to not exceed speculative byte limit and is accepted
   incorrectly.

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat commented:
> Looks great! I would love to see a unit test added since this did fix a bug.

Thanks for calling out the unit test I totally forgot. Added it as the first commit where it will fail and will pass with the remainder of the commits in the PR applied.

dotnwat commented:
Force push https://github.com/redpanda-data/redpanda/compare/3726219a993c265f81d30ab9d4291584f3e70da6..cfc74bb348b86c883efad7fb4c4acfe146b4e5b3 kept the base so the inter-diff is clean.
